### PR TITLE
Move GoodSignal from `Service.lua` to its own file

### PR DESCRIPTION
**Pull request by ccuser44**
Move GoodSignal from `Service.lua` to its own file and add license to comply with law

### CODE REVIEW

**Comment by coasterteam** [comment]
Lacks the additional backwards compatibility that was setup with the in-Service version, will review with changes later today when I get time

**Comment by ccuser44** [comment]
> Lacks the additional backwards compatibility that was setup with the in-Service version, will review with changes later today when I get time

I did a diffcheck and the only changes were the removals of `Signal:ConnectOnce`, `Signal:connectOnce`, `Signal:wait`, `Signal:fire`, `Signal:connect`, `connection:Fire`, `connection:fire`, `connection:disconnect`, `connection:wait`, `Event = {},` and `Signal:Destroy`.
I am not sure why the removal of these would make the compatibility any better so I decided to keep them.

**Comment by coasterteam** [comment]
This was due to GoodSignal being the replacement of the old wrapped BindableEvents. They had those methods existing so I didn't want to break existing events someone might have/use.

**Comment by ccuser44** [comment]
> This was due to GoodSignal being the replacement of the old wrapped BindableEvents. They had those methods existing so I didn't want to break existing events someone might have/use.

But wouldn't removal exactly break said existing events? This pull doesn't remove anything but the one in `Service.lua` does.

**Comment by coasterteam** [comment]
I modified GoodSignal with additional methods as wrapped BindableEvents were removed from the Events system. I am not sure how this is confusing?

**Comment by ccuser44** [comment]
> I modified GoodSignal with additional methods as wrapped BindableEvents were removed from the Events system. I am not sure how this is confusing?

So do I remove the added methods?

File: MainModule/Server/Shared/GoodSignal.lua [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+function Signal:Connect(fn)
+	local connection = Connection.new(self, fn)
+	if self._handlerListHead then
+		connection._next = self._handlerListHead
+		self._handlerListHead = connection
+	else
+		self._handlerListHead = connection
+	end
+	return connection
```
Comment (coasterteam) [inline MainModule/Server/Shared/GoodSignal.lua:141]:
```suggestion
    local signalSelf = self
	function connection:Fire(...)
		signalSelf:Fire(fn)
	end
	function connection:fire(...)
		signalSelf:Fire(fn)
	end
	function connection:disconnect()
		self:Disconnect()
	end
	function connection:wait()
		return signalSelf:Wait()
	end
	return connection
```

File: MainModule/Server/Shared/GoodSignal.lua [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+			if not freeRunnerThread then
+				freeRunnerThread = coroutine.create(runEventHandlerInFreeThread)
+				-- Get the freeRunnerThread to the first yield
+				coroutine.resume(freeRunnerThread)
+			end
+			task.spawn(freeRunnerThread, item._fn, ...)
+		end
+		item = item._next
+	end
+end
```
Comment (coasterteam) [inline MainModule/Server/Shared/GoodSignal.lua:167]:
```suggestion
end

function Signal:connect(fn)
    return self:Connect(fn)
end
```

File: MainModule/Server/Shared/GoodSignal.lua [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+-- a Signal:Connect() which disconnects itself.
+function Signal:Wait()
+	local waitingCoroutine = coroutine.running()
+	local cn;
+	cn = self:Connect(function(...)
+		cn:Disconnect()
+		task.spawn(waitingCoroutine, ...)
+	end)
+	return coroutine.yield()
+end
```
Comment (coasterteam) [inline MainModule/Server/Shared/GoodSignal.lua:179]:
```suggestion
end

function Signal:fire(...)
	self:Fire(...)
end
```

File: MainModule/Server/Shared/GoodSignal.lua [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+function Signal:Once(fn)
+	local cn;
+	cn = self:Connect(function(...)
+		if cn._connected then
+			cn:Disconnect()
+		end
+		fn(...)
+	end)
+	return cn
+end
```
Comment (coasterteam) [inline MainModule/Server/Shared/GoodSignal.lua:192]:
```suggestion
end

function Signal:ConnectOnce(fn)
	return self:Once(fn)
end

function Signal:connectOnce(fn)
	return self:Once(fn)
end

function Signal:Destroy()
	self:DisconnectAll()
	setmetatable(self, nil)
end
```

Comment (watameln) [reply MainModule/Server/Shared/GoodSignal.lua:192]: > 

Erm why are you removing end then adding end

Comment (coasterteam) [reply MainModule/Server/Shared/GoodSignal.lua:192]: its the same code just github review sucks on web

**Comment by coasterteam** [comment]
Added reviewed changes to ensure backwards compatibility with the events system this was replacing.

File: MainModule/Server/Shared/GoodSignal.lua (hunk 1 above)
Comment (ccuser44) [reply MainModule/Server/Shared/GoodSignal.lua:155]: Done

**Comment by ccuser44** [comment]
> Added reviewed changes to ensure backwards compatibility with the events system this was replacing.

I merged the suggestions now.

I somehow thought that the additions were actually in the vanilla GoodSignal and were removed in Adonis for compatibility but apparently it was the other way around.

Anyways it's fixed now.

**Comment by Dimenpsyonal** [comment]
Very cool

**Pull request merged**